### ISSUE TITLE
[RFC] remove 'categorical_feature' and 'feature_name' parameters in cv() and train()

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#6435]
## Proposal

I'm requesting comment on the following proposal:

* remove keyword argument `categorical_feature` from `cv()` and `train()` in the R and Python packages
* remove keyword argument `feature_name` from `cv()` and `train()` in the Python package
* remove keyword argument `colnames` from `cv()` and `train()` in the R package

And doing all of these only after the packages issuing deprecation warnings for 2-3 releases.

## Summary

Both the R and Python packages expose functions `cv()` (for cross-validation) and `train()` (for regular entire-dataset training). These functions require a LightGBM `Dataset` object.

The `Dataset` object holds attributes `categorical_features` and `feature_names`, and allows setting those via constructor keyword arguments and `set_{attr}()` methods.

Despite that, these `cv()` and `train()` functions *also* take `categorical_features` and `feature_names` as keyword arguments.

Python `cv()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L569-L570

Python `train()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L62-L63

R-package `cv()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/R-package/R/lgb.cv.R#L90-L91

R-package `train()`

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/R-package/R/lgb.train.R#L57-L58

These keyword arguments aren't providing any value, in my opinion. Their values are just forwarded along to calls like this:

https://github.com/microsoft/LightGBM/blob/92a8741b9cb36a13d7232272e7052ca5e2fa04e6/python-package/lightgbm/engine.py#L738-L740

Which at best is redundant with the `Dataset` class, and at worst could lead to runtime exceptions (if the `Dataset` has already been constructed).

## Motivation

Would simplify the library's interface without any loss of functionality.

If this proposal is accepted, the `Dataset` class would be the only place that this information is provided to `train()` and `cv()`.

## References

Inspired by this post I noticed on Stack Overflow: https://stackoverflow.com/questions/78383840/in-lightgbm-why-do-the-train-and-the-cv-apis-accept-categorical-feature-argument/78405996#78405996

`xgboost` does not expose such arguments in `train()` ([code link](https://github.com/dmlc/xgboost/blob/5e64276a9b95df57e6dd8f9e63347636f4e5d331/python-package/xgboost/training.py#L52)) or `cv()` ([code link](https://github.com/dmlc/xgboost/blob/5e64276a9b95df57e6dd8f9e63347636f4e5d331/python-package/xgboost/training.py#L410)).

These arguments have been part of the API since September 2017: https://github.com/microsoft/LightGBM/commit/ef77806934f6b020c6425372f705fd0240a68f5b#diff-9bd633ead0bdfe9540c42a618efd9e559cca16c522ad844a09fcf4ffc7d6e84c.

**Comment by borchero** [comment]
I was not aware that there parameters were even available, I'm very much in favor of the removal across language APIs.

**Comment by mayer79** [comment]
Very good idea. I have in mind that it is also in the `fit()` method of the Scikit-learn API. Do we need it there because that is the first time LightGBM sees the data?

**Comment by jameslamb** [comment]
Thanks for pointing that out @mayer79 

Yes, the `scikit-learn` interface needs these inputs because it takes in the raw data (e.g. numpy, scipy, pandas), not a `lightgbm.Dataset`. And that it by design, to match the types of inputs estimators in the `scikit-learn` library accept.

**Comment by jameslamb** [comment]
While working through this tonight to add deprecation warnings, I found a few more things in `lgb.cv()` in the R package that I think should be deprecated.

* passing anything other than a `Dataset` to `data`

https://github.com/microsoft/LightGBM/blob/6e78e697f8e27d4bbf493a0ebd9134c3254a2ecd/R-package/R/lgb.cv.R#L103-L109

* passing `label` and `weight`

https://github.com/microsoft/LightGBM/blob/6e78e697f8e27d4bbf493a0ebd9134c3254a2ecd/R-package/R/lgb.cv.R#L79-L80

`{xgboost}` does not support passing such things in `cv()` in its R or Python packages.

LightGBM's `python` package also does not support those things.

I think restricting `cv()` to only take in a `Dataset` will help simplify that interface.

cc @mayer79

**Comment by mayer79** [comment]
I like. Maybe worth noting that the one and only @david-cortes very recently removed non-dataset input in `xgb.cv()`.

**Comment by jameslamb** [comment]
Oh nice, thanks @mayer79 ! I hadn't seen https://github.com/dmlc/xgboost/pull/10031.

**Comment by StrikerRUS** [comment]
I'm +1 for all the changes discussed in this issue.

**Comment by jameslamb** [comment]
Thanks! I'll put up a PR removing these deprecated arguments. The deprecation warnings have now been part of 2 releases.
